lib/auth/auth.go

```diff
@@ -1170,6 +1170,88 @@ func (a *Server) GetMFADevices(ctx context.Context, req *proto.GetMFADevicesRequ
 	}, nil
 }
 
+// verifyMFARespAndAddDeviceWithToken validates MFA register response and on success,
+// adds the new MFA device. This method is specifically for secrets and storage keys associated with a user token.
+func (a *Server) verifyMFARespAndAddDeviceWithToken(ctx context.Context, regResp *proto.MFARegisterResponse, token types.UserToken, deviceName string) error {
+	cap, err := a.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Validate MFARegisterResponse and upsert the new device on success.
+	var dev *types.MFADevice
+	switch resp := regResp.GetResponse().(type) {
+	case *proto.MFARegisterResponse_TOTP:
+		if cap.GetSecondFactor() == constants.SecondFactorU2F {
+			return trace.BadParameter("cluster only allows U2F for second factor")
+		}
+
+		secrets, err := a.Identity.GetUserTokenSecrets(ctx, token.GetName())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		dev, err = services.NewTOTPDevice(deviceName, secrets.GetOTPKey(), a.clock.Now())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := a.checkTOTP(ctx, token.GetUser(), resp.TOTP.Code, dev); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := a.UpsertMFADevice(ctx, token.GetUser(), dev); err != nil {
+			return trace.Wrap(err)
+		}
+
+	case *proto.MFARegisterResponse_U2F:
+		if cap.GetSecondFactor() == constants.SecondFactorOTP {
+			return trace.BadParameter("cluster only allows OTP for second factor")
+		}
+
+		u2fConfig, err := cap.GetU2F()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// u2f.RegisterVerify will upsert the new device internally.
+		dev, err = u2f.RegisterVerify(ctx, u2f.RegisterVerifyParams{
+			DevName: deviceName,
+			Resp: u2f.RegisterChallengeResponse{
+				RegistrationData: resp.U2F.RegistrationData,
+				ClientData:       resp.U2F.ClientData,
+			},
+			ChallengeStorageKey:    token.GetName(),
```

Comment:
Don't we usually set both challenge and registration keys to the username?

Reply:
> as long as you have a token and the u2f config set, you can get a challenge despite the second factor setting

Isn't this a problem? Shouldn't it respect the overall MFA settings?

Reply:
not sure, it seems you can add other types of device but you won't be able to use it (so maybe not a problem?). i tested this with tsh, and if my auth setting is `u2f`, i'm still able to add a `totp` device. I'm not sure if this is by design

Reply:
i just now understood why in some places we prevent users from adding a incorrect device, `invite/reset/recovery` should prevent incorrect device otherwise they are locked out and have to be reset.

Reply:
> should prevent incorrect device otherwise they are locked out and have to be reset

That is a concern I had about this too. Do you mind covering the proper scenarios with tests, as usual?

Reply:
tested